packages/waku/src/router/client.ts

```diff
@@ -606,3 +607,244 @@ class ErrorBoundary extends Component<
+      if (staticPathSetRef.current.has(route.path)) {
+        return;
+      }
+      const skip = Array.from(cachedIdSetRef.current);
+      const rscPath = encodeRoutePath(route.path);
+      prefetchRsc(rscPath, JSON.stringify({ query: route.query, skip }));
+      (globalThis as any).__WAKU_ROUTER_PREFETCH__?.(route.path);
+    },
+    [cachedIdSetRef, staticPathSetRef],
+  );
+
+  useEffect(() => {
+    const callback = () => {
+      const route = parseRoute(new URL(window.location.href));
+      changeRoute(route);
+    };
+    window.addEventListener('popstate', callback);
+    return () => {
+      window.removeEventListener('popstate', callback);
+    };
+  }, [changeRoute]);
+
+  useEffect(() => {
+    const callback = (path: string, query: string) => {
+      const url = new URL(window.location.href);
+      url.pathname = path;
+      url.search = query;
+      url.hash = '';
+      if (path !== '/404') {
+        window.history.pushState(
+          {
+            ...window.history.state,
+            waku_new_path: url.pathname !== window.location.pathname,
+          },
+          '',
+          url,
+        );
+      }
+      changeRoute(parseRoute(url), { skipRefetch: true });
+    };
+    const listeners = (routerData[1] ||= new Set());
+    listeners.add(callback);
+    return () => {
+      listeners.delete(callback);
+    };
+  }, [changeRoute, routerData]);
+
+  useEffect(() => {
+    const { hash } = window.location;
+    const { state } = window.history;
+    const element = hash && document.getElementById(hash.slice(1));
+    window.scrollTo({
+      left: 0,
+      top: element ? element.getBoundingClientRect().top + window.scrollY : 0,
+      behavior: state?.waku_new_path ? 'instant' : 'auto',
+    });
+  });
+
+  const routeElement = createElement(Slot, { id: getRouteSlotId(route.path) });
+
+  return createElement(
+    RouterContext.Provider,
+    { value: { route, changeRoute, prefetchRoute } },
+    routeElement,
+  );
+};
+
+export function NewRouter({ routerData = DEFAULT_ROUTER_DATA }) {
+  const route = parseRouteFromLocation();
+  const initialRscPath = encodeRoutePath(route.path);
+  // FIXME cachedIdSetRef and staticPathSetRef should be initialized from the initial RSC payload
+  const cachedIdSetRef = useRef(new Set<string>());
+  const staticPathSetRef = useRef(new Set<string>());
+  const unstable_enhanceCreateData =
+    (
+      createData: (
+        responsePromise: Promise<Response>,
+      ) => Promise<Record<string, ReactNode>>,
+    ) =>
+    async (responsePromise: Promise<Response>) => {
+      const response = await responsePromise;
+      const has404 = routerData[2];
+      if (response.status === 404 && has404) {
+        // HACK this is still an experimental logic. It's very fragile.
+        // FIXME we should cache it if 404.txt is static.
+        return fetchRsc(encodeRoutePath('/404'));
+      }
+      const data = createData(responsePromise);
+      Promise.resolve(data)
```

Comment:
Does this resolve the responsePromise twice? First on L772 then again here? Or is my understanding off with that?

then if it does resolve twice, shouldn't createData just use `response`?

Reply:
createData requires a promise, so L772 isn't related with L780.

However, yes, `data.then((date) => ` here might work. The reason I added `Promise.resolve()` is because `data` returned by `createData` isn't a real promise. TS types aren't correct here. So, it's just a safe bet. Maybe it's not necessary.